g11n-ws/vip-manager-l10n/src/main/java/com/vmware/l10n/utils/S3Util.java

```diff
@@ -0,0 +1,283 @@
+
+import com.amazonaws.SdkClientException;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.amazonaws.services.s3.model.GetBucketLocationRequest;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.vmware.l10n.conf.RsaCryptUtil;
+import com.vmware.vip.common.constants.ConstantsChar;
+import com.vmware.vip.common.constants.ConstantsFile;
+import com.vmware.vip.common.constants.ConstantsKeys;
+import com.vmware.vip.common.i18n.dto.SingleComponentDTO;
+import com.vmware.vip.common.i18n.resourcefile.ResourceFilePathGetter;
+
+@Component("S3Util")
+@Profile("s3")
+public class S3Util {
+	private static Logger logger = LoggerFactory.getLogger(S3Util.class);
+
+	private AmazonS3 s3Inst;
+
+	/**
+	 * the s3 password is encryption or not
+	 */
+	@Value("${s3.password.encryption:false}")
+	private boolean encryption;
+
+	/**
+	 * the s3 password public key used to decrypt data
+	 */
+	@Value("${s3.password.publicKey}")
+	private String publicKey;
+
+	/**
+	 * the s3 access Key
+	 */
+	@Value("${s3.password.accessKey}")
+	private String accessKey;
+
+	/**
+	 * the s3 secret key
+	 */
+	@Value("${s3.password.secretkey}")
+	private String secretkey;
+
+	/**
+	 * the s3 region name
+	 */
+	@Value("${s3.region}")
+	private String s3Region;
+
+	/**
+	 * the s3 bucket Name
+	 */
+	@Value("${s3.bucketName}")
+	private String bucketName;
+
+	private Random random = new Random(System.currentTimeMillis());
+
+	private static long retryInterval = 500; // milliseconds
+	private static long deadlockInterval = 10 * 60 * 1000L; // 10 minutes
+	private static long waitS3Operation = 100; // milliseconds
+	
+	/**
+	 * initialize the the S3 client environment
+	 */
+	@PostConstruct
+	private void init() {
+		s3Inst = AmazonS3ClientBuilder.standard()
+				.withCredentials(new AWSStaticCredentialsProvider(
+						new BasicAWSCredentials(this.getAccessKey(), this.getSecretkey())))
+				.withRegion(s3Region).enablePathStyleAccess().build();
+		if (!s3Inst.doesBucketExistV2(bucketName)) {
+			s3Inst.createBucket(bucketName);
+			// Verify that the bucket was created by retrieving it and checking its
+			// location.
+			String bucketLocation = s3Inst.getBucketLocation(new GetBucketLocationRequest(bucketName));
+			logger.info("Bucket location: {}", bucketLocation);
+		}
+	}
+
+	public String readBundle(String basePath, SingleComponentDTO compDTO) {
+		logger.info("read bundle file: {}/{}/{}/{}", compDTO.getProductName(), compDTO.getVersion(),
+				compDTO.getComponent(), compDTO.getLocale());
+
+		String bundlePath = getBundleFilePath(basePath, compDTO);
+		String result = null;
+		try {
```

Comment:
If throw them all out, need to change existing code.

Reply:
It's not a good practice to catch the business exception on the lower layer, the exceptions should be on the place upper than business layer. Otherwise, the upper layer can't get what happen inside and handle it, e.g. if the upper layer wants to put the exception in the API response, there's no way to do it.

Reply:
done